README.md
simple_sentinel
===============
[![Build Status](https://travis-ci.org/vanchi-zendesk/simple_sentinel.svg?branch=v0.1.1)](https://travis-ci.org/vanchi-zendesk/simple_sentinel)

A simple nodejs utility to start and stop a test sentinel and redis.
Starts redis at port 16379 and sentinel at port 26379.

This is mainly intended as a test helper. If you have several projects which needs to run tests against redis and sentinel, you could include simple_sentinel as a package, and use it to start and stop your redis/sentinel setup.


Configuration
=============

For redis: we start redis with the following config
```
  redis-server
          --daemonize yes
          --logfile <file>
          --dir <dir>
          --loglevel verbose
          --port <port>
```
For sentinel: Created by reading this page: http://redis.io/topics/sentinel
```
  redis-server
          --sentinel
          --daemonize yes
          --logfile <file>
          --pidfile <file>
          --loglevel verbose
          --port <port>
```
with additional configs:
```
  sentinel monitor <master_name> <host> <port> 1
  sentinel down-after-milliseconds <master_name> 2000
  sentinel failover-timeout <master_name> 90000
  sentinel parallel-syncs <master_name> 2
```

Requirements
============

- redis (tested @2.8.4)
- node
- Include this package in your project, run npm install.

Usage
=====

```
-$ node_modules/simple_sentinel/bin/simple_sentinel --help

  Usage: simple_sentinel [start|stop]

  Commands:

    start     start redis and sentinel
    stop      stop redis and sentinel

  Options:

    -h, --help  output usage information
    -f, --config specify a config file
    -r, --redis <ports separated by comma>
    -s, --sentinel <ports separated by comma>
```

Config file
===========
We can specify different ports for redis and sentinels. The first redis port is treated as the master.
```
module.exports = {
  redis : {
    ports: [ 16379, 16380, 16381 ]
  },
  sentinel: {
    ports: [ 26379, 26380, 26381 ]
  }
};
```

In code
=======

You can also use simple_sentinel from code.

```
var simple_sentinel = require('simple_sentinel');
return_code = simple_sentinel.start(config);
return_code = simple_sentinel.stop(config);
```

Notes about stop
================

Stops redis:16379 by sending 'redis-cli -p 16379 shutdown', so watch out if you have redis running on that port.

For the sentinel, since the 'shutdown' command is not working on redis v2.8.4, we try to 'kill -TERM `cat sentinel.pid`'. If the pid file is not found, we further query by doing 'redis-cli -p 26379 | grep process_id', parse the pid and kill it.


## Copyright and license

Copyright 2014-2015 Zendesk

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.

You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


Changelog.md
### 0.1.8
* Add timeouts to tests to avoid failures due to slow redis start/stop
* Add an Apache V2 LICENSE file, and add a Copyright entry to README.md



test/simple_sentinel_test.js
require('shelljs/global');
var assert         = require('assert');
var SimpleSentinel = require('../index.js');

process.env.noverbose = true;
describe('simple_sentinel', function() {
  describe('without config', function() {
    it('should start redis and sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.start(null));
      assert.equal(0, exec('redis-cli -p 16379 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26379 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26379 sentinel master mymaster', {silent: true}).code);
    });

    it('should stop redis and sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.stop(null, true));
      assert.ok(0 !== exec('redis-cli -p 16379 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 26379 ping', {silent: true}).code);
    });
  });

  describe('with config', function() {
    var config = require('../sample_config.js');
    it('should start redis and sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.start(config));
      assert.equal(0, exec('redis-cli -p 16379 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 16380 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 16381 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26379 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26380 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26381 ping', {silent: true}).code);
      assert.equal(0, exec('redis-cli -p 26379 sentinel master mymaster', {silent: true}).code);
    });

    it('should stop redis and sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.stop(config));
      assert.ok(0 !== exec('redis-cli -p 16379 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 16380 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 16381 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 26379 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 26380 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 26381 ping', {silent: true}).code);
    });
  });

  describe('with singleton redis config', function() {
    var config = { redis: { ports: [ 16379 ] } };
    it('should start redis', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.start(config));
      assert.equal(0, exec('redis-cli -p 16379 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 16380 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 26379 ping', {silent: true}).code);
    });

    it('should stop redis', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.stop(config));
      assert.ok(0 !== exec('redis-cli -p 16379 ping', {silent: true}).code);
    });
  });

  describe('with no redis ports', function() {
    var config = { sentinel: { ports: [ 26379 ] } };
    it('should start sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.start(config));
      assert.equal(0, exec('redis-cli -p 26379 ping', {silent: true}).code);
      assert.ok(0 !== exec('redis-cli -p 16379 ping', {silent: true}).code);
    });

    it('should stop sentinel', function() {
      this.timeout(10000);
      assert.equal(0, SimpleSentinel.stop(config));
      assert.ok(0 !== exec('redis-cli -p 26379 ping', {silent: true}).code);
    });
  });

});


index.js
require('shelljs/global');

function redis_alive(port, verbose){
  return (exec('redis-cli -p '+port+' ping', { silent: !verbose }).code === 0);
}

function sentinel_alive(port, verbose){
  return (exec('redis-cli -p '+ port +' ping', { silent: !verbose }).code === 0);
}

function start_redis(port, slaveof) {
  var silent         = process.env.noverbose;
  var redis_dir      = __dirname + '/redis/' + port;
  var redis_log_file = redis_dir + '/redis.log';
  var redis_start    = 'redis-server ' +
                       ' --daemonize yes' +
                       ' --logfile ' + redis_log_file +
                       ' --dir ' + redis_dir +
                       ' --loglevel verbose' +
                       ' --port '+port;
  mkdir('-p', redis_dir);
  rm('-f', redis_log_file);

  if(!silent) console.log('Starting redis:'+port);
  if(!which('redis-server')) {
    console.err('Please install redis >2.8.4');
    return(1);
  }

  if(slaveof) {
    redis_start = redis_start + ' --slaveof '+slaveof;
  }

  if (exec(redis_start).code !== 0) {
    if(test('-f', redis_log_file)) {
      console.err('Failed to start redis. Here\'s the log:');
      cat(redis_log_file);
      return(1);
    }
    console.err('Failed to start redis, bailing...');
    return(1);
  }

  exec('sleep 1');

  if(!redis_alive(port)) {
    if(test('-f', redis_log_file)) {
      console.err('Redis failed ping, Here\'s the log:');
      cat(redis_log_file);
      return(1);
    }
    console.err('Redis failed ping, bailing...');
    return(1);
  }
  if(!silent) console.log('success');
}

function stop_redis(port) {
  var silent = process.env.noverbose;
  var redis_dir = __dirname + '/redis/' + port;
  var redis_log_file = redis_dir + '/redis.log';
  if(!silent) console.log('Shutting down redis:'+port);
  if(redis_alive(port)) {
    exec('redis-cli -p '+port+' shutdown');
  } else {
    if(!silent) console.log('already down');
  }
  rm('-f', redis_log_file);
  rm('-rf', redis_dir);
}

function build_sentinel_config(name, host, port) {
  if(name && host && port) {
    return [ 'sentinel monitor '+name+ ' '+host+' '+port+' 1',
             'sentinel down-after-milliseconds '+name+' 2000',
             'sentinel failover-timeout '+name+' 90000',
             'sentinel parallel-syncs '+name+' 2'].join('\n');
  } else {
    return "";
  }
}

function start_sentinel(port, master_host, master_port) {
  var silent             = process.env.noverbose;
  var sentinel_dir       = __dirname + '/sentinel/' + port;
  var sentinel_conf_path = __dirname + '/sentinel/' + port + '/sentinel.conf';
  var sentinel_log_file  = sentinel_dir + '/sentinel.log';
  var sentinel_pid_file  = sentinel_dir + '/sentinel.pid';
  var sentinel_start     = 'redis-server ' + sentinel_conf_path +
                           ' --sentinel ' +
                           ' --daemonize yes' +
                           ' --logfile ' + sentinel_log_file +
                           ' --pidfile ' + sentinel_pid_file +
                           ' --port ' + port +
                           ' --loglevel verbose';
  mkdir('-p', sentinel_dir);
  rm('-f', sentinel_log_file);

  var fs = require('fs');
  var fd = fs.openSync(sentinel_conf_path, 'w');
  fs.writeSync(fd, build_sentinel_config('mymaster', master_host, master_port));
  fs.closeSync(fd);

  if(!silent) console.log('Starting sentinel:'+port);
  if(exec(sentinel_start).code !== 0) {
    if(test('-f', sentinel_log_file)) {
      console.err('Failed to start sentinel. Here\'s the log:');
      cat(sentinel_log_file);
      return(1);
    }
    console.err('Failed to start sentinel, bailing...');
    return(1);
  }

  exec('sleep 1');

  if(!sentinel_alive(port)) {
    if(test('-f', sentinel_log_file)) {
      console.err('Sentinel failed ping. Here\'s the log:');
      cat(sentinel_log_file);
      return(1);
    }
    console.err('Sentinel failed ping, bailing...');
    return(1);
  }
  if (!silent) console.log('success');
}

function stop_sentinel(port) {
  var silent             = process.env.noverbose;
  var sentinel_dir       = __dirname + '/sentinel/' + port;
  var sentinel_log_file  = __dirname + '/sentinel/'+ port +'/sentinel.log';
  var sentinel_pid_file  = __dirname + '/sentinel/'+ port +'/sentinel.pid';
  var sentinel_conf_path = __dirname + '/sentinel/' + port + '/sentinel.conf';
  if(!silent) console.log('Shutting down sentinel:'+port);
  if(sentinel_alive(port)) {
    var pid;

    if(test('-f', sentinel_pid_file)) {
      pid = exec('cat '+sentinel_pid_file, {silent:true}).output;
    } else {
      pid = exec('redis-cli -p '+ port +' info | grep process_id | cut -d: -f2', {silent:true}).output;
    }
    if(pid) {
      pid = parseInt(pid);
      exec('kill -TERM ' + pid);
    }
  } else {
    if(!silent) console.log('already down');
  }
  rm('-f', sentinel_log_file);
  rm('-f', sentinel_conf_path);
  rm('-rf', sentinel_dir);
}

var default_config = {
  redis : {
    ports: [ 16379 ]
  },
  sentinel: {
    ports: [ 26379 ]
  }
};

module.exports = {

  start: function(config) {
    var i;
    config = config || default_config;
    if(config.redis && config.redis.ports) {
      var slaveof = 'localhost '+config.redis.ports[0];
      for(i = 0; i < config.redis.ports.length; i++) {
        start_redis(config.redis.ports[i], (i === 0)?'':slaveof);
      }
    }
    if(config.sentinel && config.sentinel.ports) {
      for(i = 0; i < config.sentinel.ports.length; i++) {
        start_sentinel(config.sentinel.ports[i], 'localhost', (config.redis && config.redis.ports && config.redis.ports[0]));
      }
    }
    return(0);
  },

  stop: function(config) {
    config = config || default_config;
    var i;
    if(config.redis && config.redis.ports) {
      var slaveof = 'localhost '+config.redis.ports[0];
      for(i = 0; i < config.redis.ports.length; i++) {
        stop_redis(config.redis.ports[i]);
      }
    }
    if(config.sentinel && config.sentinel.ports) {
      for(i = 0; i < config.sentinel.ports.length; i++) {
        stop_sentinel(config.sentinel.ports[i]);
      }
    }
    return(0);
  }
};
